File: ios/Positive Only Social/Positive Only Social/uploader/AWSManager.swift
//
//  AWSManager.swift
//  Positive Only Social
//
//  Created by Andrew Katson on 10/11/25.
//

import Foundation
import AWSClientRuntime
import AWSS3
import AWSSDKIdentity
import AWSCognitoIdentity
import AWSCognitoIdentityProvider
import AwsCommonRuntimeKit
import UIKit

public enum AWSManagerError: Error, LocalizedError {
    /// The client failed to initialize during app launch.
    /// The associated error contains the reason (e.g., network, bad credentials).
    case initializationFailed(Error)
    
    /// An attempt was made to use the client, but it was not initialized.
    case clientNotInitialized
    
    public var errorDescription: String? {
        switch self {
        case .initializationFailed(let error):
            return "AWSManager failed to initialize: \(error.localizedDescription)"
        case .clientNotInitialized:
            return "AWS S3 Client is not initialized. Check logs for a detailed error."
        }
    }
}

// RECOMMENDED: This implementation uses the modern AWS SDK for Swift (v3)
final class AWSManager {
    static let shared = AWSManager()

    /// Publicly expose the S3 client for use in other parts of the app (like S3Uploader)
    private(set) var s3Client: S3Client?
    
    private(set) var initializationError: Error?
    
    /// Publicly expose the region for URL construction
    let awsRegion = "us-east-2"
    private let identityPoolId = "us-east-2:445cf6ff-6f59-4cff-94c9-51db170ad81e"

    private init() {
        do {
            let credentialsResolver = try CognitoAWSCredentialIdentityResolver(identityPoolId: identityPoolId)
            
            let configuration = try S3Client.Config(awsCredentialIdentityResolver: credentialsResolver, region: awsRegion)
            
            // Initialize the S3 client
            self.s3Client = S3Client(config: configuration)
            self.initializationError = nil
            
            // A log message is very helpful for debugging
            // We use NSLog() instead of a print() because it shows local and in Firebase.
            //See https://shorturl.at/KF2XX for more info.
            NSLog("✅ AWSManager: S3Client initialized successfully.")
            
        } catch {
            // 3. This is the new graceful handling.
            // Instead of crashing, we log the error and set our client to nil.
            
            NSLog("❌ AWSManager: Failed to initialize AWS S3 Client: \(error)")
            self.s3Client = nil
            self.initializationError = AWSManagerError.initializationFailed(error)
        }
    }
}

final class S3Uploader {
    private let s3Client = AWSManager.shared.s3Client
    private let bucketName = "goodvibesonly-images"

    /// Uploads data to S3 and returns the public URL.
    func upload(data: Data, fileName: String) async throws -> URL {
        
        let compressedData = self.compressImage(data:data,maxSizeBytes: 10 * 1024 * 1024) //10 MB
        
        let input = PutObjectInput(
            body: .data(compressedData),
            bucket: bucketName,
            contentType: "image/jpeg",
            key: fileName
        )
        
       
        guard let s3Client = self.s3Client else {
            // If the client is nil, throw our custom error.
            // The call site (e.g., your ViewModel) can now catch this.
            throw AWSManagerError.clientNotInitialized
        }

        _ = try await s3Client.putObject(input: input)

        // Construct the public URL for the uploaded object
        let region = AWSManager.shared.awsRegion
        let urlString = "https://\(bucketName).s3.\(region).amazonaws.com/\(fileName)"
        guard let url = URL(string: urlString) else {
            throw URLError(.badURL)
        }

        NSLog("✅ Successfully uploaded to S3. URL: \(url)")
        return url
    }
        
    /**
         * Compresses the image data to be within the specified max size.
         * @param data The original image data
         * @param maxSizeBytes The maximum allowed size in bytes
         * @return The compressed image data
         */
    func compressImage(data: Data, maxSizeBytes: Int) -> Data {
        // 1. Check if already within limits
        if data.count <= maxSizeBytes {
            NSLog("✅ Good news ,Image size (\(data.count) bytes) is within limits.")
            return data
        }
        
        // 2. Convert Data to UIImage
        guard let image = UIImage(data: data) else {
            NSLog("❌ Oops,Error: Could not decode image data")
            return data // Return original data if conversion fails
        }

        // 3. Compress using JPEG (PNG does not support compression levels)
        var compression: CGFloat = 0.9
        guard var imageData = image.jpegData(compressionQuality: compression) else {
            return data
        }

        // 4. Iteratively reduce quality until size is met
        while imageData.count > maxSizeBytes && compression > 0.1 {
            compression -= 0.1
            if let compressedImage = image.jpegData(compressionQuality: compression) {
                imageData = compressedImage
            }
        }

        // 5. If still too large at minimum quality, progressively downscale the image
        if imageData.count > maxSizeBytes {
            var currentImage = image
            var currentSize = currentImage.size
            
            // Safety: avoid infinite loops by enforcing a minimum dimension
            let minimumDimension: CGFloat = 1.0
            let scaleFactor: CGFloat = 0.9
            
            while imageData.count > maxSizeBytes &&
                  currentSize.width > minimumDimension &&
                  currentSize.height > minimumDimension {
                
                currentSize = CGSize(width: currentSize.width * scaleFactor,
                                     height: currentSize.height * scaleFactor)
                
                UIGraphicsBeginImageContextWithOptions(currentSize, false, currentImage.scale)
                currentImage.draw(in: CGRect(origin: .zero, size: currentSize))
                let resizedImage = UIGraphicsGetImageFromCurrentImageContext()
                UIGraphicsEndImageContext()
                
                guard let resized = resizedImage,
                      let resizedData = resized.jpegData(compressionQuality: compression) else {
                    // If resizing or encoding fails, break and use the best effort so far
                    break
                }
                
                currentImage = resized
                imageData = resizedData
            }
        }
        return imageData
    }
}

File: android/PositiveOnlySocial/app/src/main/java/com/example/positiveonlysocial/data/uploader/AWSManager.kt
package com.example.positiveonlysocial.data.uploader

import android.util.Log
import aws.sdk.kotlin.services.cognitoidentity.CognitoIdentityClient
import aws.sdk.kotlin.services.cognitoidentity.getCredentialsForIdentity
import aws.sdk.kotlin.services.cognitoidentity.getId
import aws.sdk.kotlin.services.s3.S3Client
import aws.sdk.kotlin.services.s3.model.PutObjectRequest
import aws.smithy.kotlin.runtime.auth.awscredentials.Credentials
import aws.smithy.kotlin.runtime.auth.awscredentials.CredentialsProvider
import aws.smithy.kotlin.runtime.collections.Attributes
import aws.smithy.kotlin.runtime.content.ByteStream
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

// Custom Exception Class
sealed class AWSManagerError(message: String, cause: Throwable? = null) : Exception(message, cause) {

    class InitializationFailed(cause: Throwable) :
        AWSManagerError("AWSManager failed to initialize: ${cause.localizedMessage}", cause)

    object ClientNotInitialized :
        AWSManagerError("AWS S3 Client is not initialized. Check logs for a detailed error.") {
        private fun readResolve(): Any = ClientNotInitialized
    }
}

/**
 * RECOMMENDED: This implementation uses the modern AWS SDK for Kotlin
 */
object AWSManager {
    // Publicly expose the region
    const val AWS_REGION = "us-east-2"
    private const val IDENTITY_POOL_ID = "us-east-2:445cf6ff-6f59-4cff-94c9-51db170ad81e"

    var s3Client: S3Client? = null
        private set

    var initializationError: Throwable? = null
        private set

    /**
     * In Kotlin, network operations (like setting up credentials) are suspending.
     * Call this method from your Application class or main Activity's onCreate
     * within a CoroutineScope.
     */
    fun initialize() {
        try {
            // Create a Cognito Credentials Provider
            val cognitoProvider = CognitoCredentialsProvider(
                identityPoolId = IDENTITY_POOL_ID,
                region = AWS_REGION
            )

            // Initialize the S3 Client
            s3Client = S3Client {
                region = AWS_REGION
                credentialsProvider = cognitoProvider
            }

            initializationError = null
            Log.d("AWSManager", "✅ AWSManager: S3Client initialized successfully.")

        } catch (e: Exception) {
            // Graceful handling
            Log.e("AWSManager", "❌ AWSManager: Failed to initialize AWS S3 Client: ${e.localizedMessage}")
            s3Client = null
            initializationError = AWSManagerError.InitializationFailed(e)
        }
    }
}

class S3Uploader {
    private val bucketName = "goodvibesonly-images"

    /**
     * Uploads data to S3 and returns the public URL.
     * @param data The file bytes to upload
     * @param fileName The destination file name
     */
    suspend fun upload(data: ByteArray, fileName: String): URL = withContext(Dispatchers.IO) {

        val s3Client = AWSManager.s3Client ?: run {
            // If the client is null, throw our custom error.
            throw AWSManagerError.ClientNotInitialized
        }

        val compressedData = compressImage(data, 10 * 1024 * 1024) // 10MB

        val input = PutObjectRequest {
            bucket = bucketName
            key = fileName
            body = ByteStream.fromBytes(compressedData)
            contentType = "image/jpeg"
        }

        // Perform the upload
        s3Client.putObject(input)

        // Construct the public URL
        val region = AWSManager.AWS_REGION
        val urlString = "https://$bucketName.s3.$region.amazonaws.com/$fileName"

        try {
            val url = URL(urlString)
            Log.d("S3Uploader", "Successfully uploaded to S3. URL: $url")
            url
        } catch (e: Exception) {
            throw AWSManagerError.InitializationFailed(e) // Or a specific URLError
        }
    }

    private fun isJpeg(data: ByteArray): Boolean {
        return data.size >= 2 && data[0] == 0xFF.toByte() && data[1] == 0xD8.toByte()
    }

    private fun rotateImageIfRequired(data: ByteArray, bitmap: Bitmap): Bitmap {
        try {
            val exifInterface = ExifInterface(ByteArrayInputStream(data))
            val orientation = exifInterface.getAttributeInt(
                ExifInterface.TAG_ORIENTATION,
                ExifInterface.ORIENTATION_NORMAL
            )
            val matrix = Matrix()
            when (orientation) {
                ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
                ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
                ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
                else -> return bitmap
            }
            val rotatedBitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
            bitmap.recycle()
            return rotatedBitmap
        } catch (e: Exception) {
            Log.w("S3Uploader", "Failed to check EXIF orientation: ${e.localizedMessage}")
            return bitmap
        }
    }

    /**
     * Compresses the image data to be within the specified max size.
     * @param data The original image data
     * @param maxSizeBytes The maximum allowed size in bytes
     * @return The compressed image data
     */
    private fun compressImage(data: ByteArray, maxSizeBytes: Long): ByteArray {
        if (data.size <= maxSizeBytes && isJpeg(data)) {
            Log.d("S3Uploader", "Image is already JPEG and size (${data.size} bytes) is within limits.")
            return data
        }

        Log.d("S3Uploader", "Transcoding/compressing image to JPEG... Original size: ${data.size} bytes")

        var bitmap = BitmapFactory.decodeByteArray(data, 0, data.size) ?: run {
            Log.e("S3Uploader", "Failed to decode bitmap for compression.")
            return data
        }

        bitmap = rotateImageIfRequired(data, bitmap)

        var quality = 90
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, stream)
        var compressedData = stream.toByteArray()

        // Iteratively reduce quality until size is under limit or quality is too low
        while (compressedData.size > maxSizeBytes && quality >= 10) {
            val nextStream = ByteArrayOutputStream()
            quality -= 10
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, nextStream)
            compressedData = nextStream.toByteArray()
            Log.d("S3Uploader", "Compressed to quality $quality, size: ${compressedData.size} bytes")
        }

        bitmap.recycle() // Free memory
        return compressedData
    }
}

/**
 * Helper class to bridge Cognito Identity to the AWS Credentials Provider interface.
 * The AWS Kotlin SDK does not have a one-line "CognitoAWSCredentialIdentityResolver"
 * like Swift, so we define a simple provider.
 */
class CognitoCredentialsProvider(
    private val identityPoolId: String,
    private val region: String
) : CredentialsProvider {

    override suspend fun resolve(attributes: Attributes): Credentials = withContext(Dispatchers.IO) {
        // The AWS Kotlin SDK requires an explicit region — on Android there is no
        // ambient region (no env vars / instance metadata), so omitting it makes
        // the first call throw a region-resolution error before any request is
        // sent, which surfaced as posts "failing immediately" (issue #292).
        val identityClient = CognitoIdentityClient {
            region = this@CognitoCredentialsProvider.region
        }
        identityClient.use { client ->
            // 1. Get ID
            val idResponse = client.getId {
                identityPoolId = this@CognitoCredentialsProvider.identityPoolId
            }

            // 2. Get Credentials for ID
            val credsResponse = client.getCredentialsForIdentity {
                identityId = idResponse.identityId
            }

            val credentials = credsResponse.credentials ?: throw Exception("No credentials returned")

            Credentials(
                accessKeyId = credentials.accessKeyId ?: "",
                secretAccessKey = credentials.secretKey ?: "",
                sessionToken = credentials.sessionToken,
                expiration = credentials.expiration
            )
        }
    }
}
